drv/cosmo-hf/src/hf.rs

```diff
@@ -0,0 +1,209 @@
+        self.flash_page_program(
+            addr,
+            LeaseBufReader::<_, 32>::from(data.into_inner()),
+        )
+        .map_err(|()| RequestError::went_away())
+    }
+
+    fn read(
+        &mut self,
+        _: &RecvMessage,
+        offset: u32,
+        dest: LenLimit<Leased<W, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        self.flash_read(
+            offset,
+            LeaseBufWriter::<_, 32>::from(dest.into_inner()),
+        )
+        .map_err(|_| RequestError::went_away())
+    }
+
+    fn sector_erase(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr as usize / SECTOR_SIZE_BYTES == 0
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_sector_erase(addr);
+        Ok(())
+    }
+
+    fn get_mux(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<HfMuxState, RequestError<HfError>> {
+        todo!()
+    }
+
+    fn set_mux(
+        &mut self,
+        _: &RecvMessage,
+        _state: HfMuxState,
+    ) -> Result<(), RequestError<HfError>> {
+        todo!()
+    }
+
+    fn get_dev(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<HfDevSelect, RequestError<HfError>> {
+        todo!()
+    }
+
+    fn set_dev(
+        &mut self,
+        _: &RecvMessage,
+        _state: HfDevSelect,
+    ) -> Result<(), RequestError<HfError>> {
+        todo!()
+    }
+
+    fn hash(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        len: u32,
+    ) -> Result<[u8; SHA256_SZ], RequestError<HfError>> {
+        let hash_driver = drv_hash_api::Hash::from(HASH.get_task_id());
+        if hash_driver.init_sha256().is_err() {
+            return Err(HfError::HashError.into());
+        }
+        let begin = addr as usize;
+        // TODO: Begin may be an address beyond physical end of
+        // flash part and may wrap around.
+        let end = match begin.checked_add(len as usize) {
+            Some(end) => {
+                // Check end > maximum 4-byte address.
+                // TODO: End may be beyond physical end of flash part.
+                //       Use that limit rather than maximum 4-byte address.
+                if end > u32::MAX as usize {
+                    return Err(HfError::HashBadRange.into());
+                } else {
+                    end
+                }
+            }
+            None => {
+                return Err(HfError::HashBadRange.into());
+            }
+        };
+        // If we knew the flash part size, we'd check against those limits.
+        let mut buf = [0u8; PAGE_SIZE_BYTES];
+        for addr in (begin..end).step_by(buf.len()) {
+            let size = (end - addr).min(buf.len());
+            self.flash_read(addr as u32, &mut buf[..size]).unwrap();
```

Comment:
Tracing the data flow, this unwrap is a client panic multiplier.

`flash_read` only fails if the client-supplied borrow writer returns an error, which happens if the buffer provided is e.g. too small or otherwise violates some assumption. The length of the buffer is checked, so the only practical way the buffer operation could fail is if the client were restarted during the IPC.

This is generally hard to achieve, but can happen in practice through supervisor action, and there's no reason to kill the flash driver in response to it. This unwrap should be removed.

Reply:
`flash_read` is generic across `idol_runtime::BufWriter`, so this call is writing to a local buffer (`&mut buf[..size]`), **not** a lease.  I believe this means that `BufWriter` can't fail, so this is actually fine (but agreed that a comment would be good here).

Reply:
I see. Yeah, a comment would be good, and on the off chance that this generates a panic site, please use `unwrap_lite`.

Reply:
commented and `unwrap_lite`'d